File: python/sorting_algorithms/quick_sort_random.py
import random

def partition_random(arr, start, end):
    random_pivot = random.randrange(start, end)

    arr[start], arr[random_pivot] = arr[random_pivot], arr[start]
    print(f'arr: {arr}, pivot: {arr[0]}, random_pivot_index: {random_pivot}')
    
    return partition(arr, start, end)


def partition(arr, start, end):
    pivot = arr[start]
    swap_index =  start

    for i in range(start + 1, end + 1):
        if arr[i] <= pivot:
            swap_index += 1
            arr[swap_index], arr[i] = arr[i], arr[swap_index] 
            #swap_index += 1

    arr[start], arr[swap_index] = arr[swap_index], arr[start]

    return swap_index

def quick_sort(arr, start, end):
    if start < end:
        pivot_index = partition_random(arr, start, end)
        #pivot_index = partition(arr, start, end)
        quick_sort(arr, start, pivot_index - 1)
        quick_sort(arr, pivot_index + 1, end)
    return arr


#arr1 = [23, 41, 4, 56, 8, 98, 12, -10, 9, -12]
#arr1 = [23, 4, -10]
#arr1 = [10, 9, 8, 7, 6, 5, 4, 3, 2, 1,]
arr1 = [1,2,3,4,5,6,7,8,9,10]

print(f'{quick_sort(arr1, 0, len(arr1) - 1)}')

File: javascript/sorting_algorithms/quickSortRandom.js
const swap = (arr, indx1, indx2) => {
    [arr[indx1],arr[indx2]] = [arr[indx2],arr[indx1]];
}

function randomPivotIndx(arr, start, end) {
    let randomPivot = Math.floor(Math.random() * (end - start + 1) + start);
    swap(arr, start, randomPivot);
    return partition(arr, start, end);
}

function partition(arr, start, end) {
    let pivot = arr[start];
    let swapIndex = start + 1;
    for (let i = start + 1; i <= arr.length + 1; i++) {
        if (pivot >= arr[i]) {
            swap(arr, i, swapIndex);
            swapIndex++;
        }
    }
    swap(arr, start, swapIndex - 1);
    return swapIndex - 1;
}

function quickSort(arr, start, end) {
    if (start < end) {
        let pivotIndex = randomPivotIndx(arr, start, end);
        quickSort(arr, start, pivotIndex - 1);
        quickSort(arr, pivotIndex + 1, end);
    }
    return arr;
}

let arr1 = [12, -2, 16, 22, 65, -4, 19, 34];
console.log(quickSort(arr1, 0, arr1.length - 1));